core/src/main/java/org/infinispan/util/concurrent/locks/impl/DefaultLockManager.java

```diff
@@ -0,0 +1,395 @@
+         for (LockPromise lockPromise : lockPromiseList) {
+            if (!lockPromise.isAvailable()) {
+               return false;
+            }
+         }
+         return true;
+      }
+
+      @Override
+      public void lock() throws InterruptedException, TimeoutException {
+         Exception exception = null;
+         ExceptionType exceptionType = ExceptionType.NONE;
+         for (ExtendedLockPromise lockPromise : lockPromiseList) {
+            try {
+               //we still need to invoke lock in all the locks.
+               lockPromise.lock();
+            } catch (InterruptedException e) {
+               if (exception == null) {
+                  exception = e;
+                  exceptionType = ExceptionType.INTERRUPTED;
+               }
+            } catch (TimeoutException e) {
+               if (exception == null) {
+                  exception = e;
+                  exceptionType = ExceptionType.TIMEOUT;
+               }
+            } catch (DeadlockDetectedException e) {
+               if (exception == null) {
+                  exception = e;
+                  exceptionType = ExceptionType.DEADLOCK;
+               }
+            } catch (RuntimeException e) {
+               if (exception == null) {
+                  exception = e;
+                  exceptionType = ExceptionType.RUNTIME;
+               }
+            }
+         }
+         if (exception != null) {
+            switch (exceptionType) {
+               case INTERRUPTED:
+                  throw (InterruptedException) exception;
+               case TIMEOUT:
+                  throw (TimeoutException) exception;
+               case RUNTIME:
+                  throw (RuntimeException) exception;
+               case DEADLOCK:
+                  throw (DeadlockDetectedException) exception;
+               default:
+                  break;
+            }
+         }
+      }
+
+      @Override
+      public void addListener(LockListener listener) {
+         notifier.thenRun(() -> listener.onEvent(lockState));
+      }
+
+      @Override
+      public void setDeadlockChecker(DeadlockChecker deadlockChecker) {
+         for (LockPromise lockPromise : lockPromiseList) {
+            lockPromise.setDeadlockChecker(deadlockChecker);
+         }
+      }
+
+      @Override
+      public void onEvent(LockState state) {
+         if (state != LockState.AVAILABLE && UPDATER.compareAndSet(this, LockState.AVAILABLE, state)) {
+            for (ExtendedLockPromise lockPromise : lockPromiseList) {
+               lockPromise.cancel(state);
+            }
+         }
+         if (isAvailable()) {
+            notifier.complete(null);
+         }
+      }
+
+      @Override
+      public void addListener(KeyAwareLockListener listener) {
+         for (KeyAwareExtendedLockPromise lockPromise : lockPromiseList) {
+            lockPromise.addListener(listener);
+         }
+      }
+
+      @Override
+      public void run() {
```

Comment:
`ScheduledExecutorFactory` will create a `Callable` wrapper for `Runnable` tasks, so in this case it's better to implement `Callable`.